FLEDGE_implementation_overview.md

```diff
@@ -0,0 +1,64 @@
+
+
+## Phase 2 Bidding and Scoring Phase
+
+### Starting Worklet Executors
+Once all interest groups are loaded, Chrome starts loading the seller worklet, and all bidder worklets participating in an auction.  To load a worklet, Chrome first checks if there is already a matching worklet “executor” (Same worklet origin, same javascript URL, wasm URL, trusted signals URL, same frame - frame match is required by the DevTools hooks. "Frame" here means the frame runAdAuction() was called in.  It may be for a different auction / component auction, however).  If so, the executor is reused.  Note that different executors do not currently share any network fetches or Javascript contexts, although they do share the same global HTTP cache, so it’s generally a good idea to try and use the same URLs for all interest groups, to maximize executor reuse.  Also, interest groups that share an executor are guaranteed to be run in the priority order they were sorted into (this may change if trusted signals fetches are split up, though).  Interest groups that don’t share an executor are run on the basis of whichever has everything it needs to run first, runs first.
+
+Otherwise, Chrome requests a bidder or seller worklet process for the associated buyer/seller origin.  Buyers and sellers do not share processes, even if the origin is the same, though a single buyer or seller origin will globally use a single process shared by all of its executors. There are separate Chrome-wide limits on the number of buyer processes (10) and seller processes (3).  These numbers are not optimized, and subject to change.  On mobile, Chrome will generally reuse the single renderer process, instead of using more secure, isolated processes, due to resource constraints.  Within a process, all Protect Audience worklets use a single Javascript thread.  Once a process is assigned, Chrome creates an executor for the worklet, which starts loading the javascript and wasm URLs.  Because bidders share processes based on origin, this process means that all of a single bidder origin’s interest groups are assigned an executor (newly created or reused) at once.
```

Comment:
We use two threads (bidder and seller) in the mobile case.

re: everyone getting assigned at once: we do it in chunks and return to event loop in between; that's likely beyond the abstraction level of this document, but I am not 100% sure it doesn't affect some interleavings.

Reply:
Updated to mention bidder vs seller threads.

What chunks are you referring to?

Reply:
This stuff: https://source.chromium.org/chromium/chromium/src/+/main:content/browser/interest_group/auction_worklet_manager.cc;drc=f086e845dd8a2a49f67221007f7ce9ed535a5711;l=301

Reply:
Nice memory!  I had completely forgotten about that stuff.

I don't think it affects interleaving of a single BidderWorklet, since it keeps a FIFO queue.  If there are two BidderWorklets for a single bidder, though, it could affect the interleaving between the two worklets (particularly if there's no trusted bidding signals URL, and the JS is cached, or something).

I think I don't actually understand the implications of this well enough to talk about it.